pygen/pygen_src/riscv_illegal_instr.py

```diff
@@ -0,0 +1,368 @@
+
+"""
+Copyright 2020 Google LLC
+Copyright 2020 PerfectVIPs Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+"""
+
+import vsc
+import logging
+from enum import IntEnum, auto
+from importlib import import_module
+from pygen_src.riscv_instr_gen_config import cfg
+from pygen_src.riscv_instr_pkg import riscv_instr_group_t
+rcs = import_module("pygen_src.target." + cfg.argv.target + ".riscv_core_setting")
+
+
+# ---------------------------------------------------------------------------------------------
+# This class is used to generate illegal or HINT instructions.
+# The illegal instruction will be generated in binary format and mixed with other valid instr.
+# The mixed instruction stream will be stored in data section and loaded to instruction pages
+# at run time.
+# ---------------------------------------------------------------------------------------------
+
+
+class illegal_instr_type_e(IntEnum):
+    kIllegalOpcode = 0
+    kIllegalCompressedOpcode = auto()
+    kIllegalFunc3 = auto()
+    kIllegalFunc7 = auto()
+    kReservedCompressedInstr = auto()
+    kHintInstr = auto()
+    kIllegalSystemInstr = auto()
+
+
+class reserved_c_instr_e(IntEnum):
+    kIllegalCompressed = 0
+    kReservedAddispn = auto()
+    kReservedAddiw = auto()
+    kReservedAddi16sp = auto()
+    kReservedLui = auto()
+    kReservedLwsp = auto()
+    kReservedLdsp = auto()
+    kReservedLqsp = auto()
+    kReservedJr = auto()
+    kReservedC0 = auto()
+    kReservedC1 = auto()
+    kReservedC2 = auto()
+
+
+@vsc.randobj
+class riscv_illegal_instr:
+    def __init__(self):
+        self.comment = ""
+        self.exception = vsc.rand_enum_t(illegal_instr_type_e)
+        self.reserved_c = vsc.rand_enum_t(reserved_c_instr_e)
+        self.instr_bin = vsc.rand_bit_t(32)
+        self.opcode = vsc.rand_bit_t(7)
+        self.compressed = vsc.rand_bit_t(1)
+        self.func3 = vsc.rand_bit_t(3)
+        self.func7 = vsc.rand_bit_t(7)
+        self.has_func3 = vsc.rand_bit_t(1)
+        self.has_func7 = vsc.rand_bit_t(1)
+        self.c_op = vsc.rand_bit_t(2)
+        self.c_msb = vsc.rand_bit_t(3)
+        self.csrs = []
+        # Default legal self.opcode for RV32I instructions
+        self.legal_opcode = vsc.list_t(vsc.bit_t(7))
```

Comment:
maybe I'm not too familiar with the PyVCS syntax...is there a way to do a one-line list initialization here?

Reply:
one needs to define the type first to assign the corresponding values.
There is no way to do this in one line.